File: swift/1710_MaximumUnitsOnATruck.swift
class Solution {
    func maximumUnits(_ boxTypes: [[Int]], _ truckSize: Int) -> Int {
        var boxes = boxTypes.sorted() { $0[1] > $1[1] }
        var totalUnits = 0, currentSize = truckSize, idx = 0, N = boxes.count
        while idx < N && currentSize > 0 {
            var box = boxes[idx]
            totalUnits += (currentSize > box[0] ? box[0] : currentSize) * box [1]
            currentSize -= box[0]
            idx += 1
        }
        return totalUnits
    }
}

File: java/1710_MaximumUnitsOnATruck.java
class Solution {
public
    int maximumUnits(int[][] boxTypes, int truckSize) {
        PriorityQueue<int[]> queue = new PriorityQueue<int[]>((a, b)->b[1] - a[1]);
        queue.addAll(Arrays.asList(boxTypes));
        int unitCount = 0;
        while (truckSize > 0 && !queue.isEmpty()) {
            int[] top = queue.poll();
            int boxCount = Math.min(truckSize, top[0]);
            unitCount += boxCount * top[1];
            truckSize -= boxCount;
        }
        return unitCount;
    }
}
